internal/transport/http2_server.go

```diff
@@ -412,22 +414,22 @@ func (t *http2Server) operateHeaders(frame *http2.MetaHeadersFrame, handle func(
 			timeoutSet = true
 			var err error
 			if timeout, err = decodeTimeout(hf.Value); err != nil {
-				headerError = true
+				headerError = status.Newf(codes.Internal, "malformed grpc-timeout: %v", err)
 			}
 		// "Transports must consider requests containing the Connection header
 		// as malformed." - A41
 		case "connection":
 			if logger.V(logLevel) {
 				logger.Errorf("transport: http2Server.operateHeaders parsed a :connection header which makes a request malformed as per the HTTP/2 spec")
 			}
-			headerError = true
+			protocolError = true
 		default:
 			if isReservedHeader(hf.Name) && !isWhitelistedHeader(hf.Name) {
 				break
 			}
 			v, err := decodeMetadataHeader(hf.Name, hf.Value)
 			if err != nil {
-				headerError = true
+				headerError = status.Newf(codes.Internal, "malformed binary metadata %q in header %q: %v", hf.Value, hf.Name, err)
```

Comment:
The comment on `Internal` says the following:

```golang
	// Internal errors. Means some invariants expected by underlying
	// system has been broken. If you see one of these errors,
	// something is very broken.
```

@dfawley : Is a different code more appropriate here. I thought `InvalidArgument` or `FailedPrecondition` might be better, but both are never returned by the gRPC framework. Any thoughts?

Reply:
FWIW [this](https://github.com/grpc/grpc/blob/master/doc/statuscodes.md) is where we document what we don't generate (and includes both of your initial ideas) and what we should generate for some specific conditions.  From that doc, some other parsing-type errors are `Internal`, so I think that's appropriate for this, too.

Reply:
Thanks for that doc. It definitely has more details that I've found elsewhere. It might be worth adding a link to it from our `codes.go` file.